### PR TITLE
Add option to manage the default storage class

### LINKED ISSUE #122: Add option to enable encryption to StorageClassses
**How to categorize this issue?**

/area security
/kind enhancement
/priority normal
/platform aws

**What would you like to be added**:

Add option to configure the `StorageClasses` with `enrypted: "true"` option

**Why is this needed**:

Security

### PR DESCRIPTION
**Pull request by mvladev**
**How to categorize this PR?**

/area storage
/kind enhancement
/priority normal
/platform aws

**What this PR does / why we need it**:

New `storage.managedDefaultClass` option in the `ControlPlaneConfig` can be used enable/disable marking the `default` Storage/Volume Snapshot classes as default. This allows end-users to set the storage class to another one.

By default its set to `true` to keep the old behaviour.

**Which issue(s) this PR fixes**:

Fixes #122

**Special notes for your reviewer**:

This PR is the worked version of #123. The old PR is kept as reference.

**Release note**:

```feature user
New `storage.managedDefaultClass` in `ControlPlaneConfig` controls if the `default` storage / volume snapshot classes are marked as default by Gardener. Set it to `false` to [mark another storage / volume snapshot class as default](https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/) without Gardener overwriting this change. If unset, this field defaults to `true`.
```

### CODE REVIEW

File: pkg/apis/aws/types_controlplane.go [hunk 1]
```diff
@@ -26,10 +26,21 @@ type ControlPlaneConfig struct {
 
 	// CloudControllerManager contains configuration settings for the cloud-controller-manager.
 	CloudControllerManager *CloudControllerManagerConfig
+
+	// Storage contains configuration for storage in the cluster.
+	Storage *Storage `json:"storage,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/aws/types_controlplane.go:31]:
```suggestion
	Storage *Storage
```

File: pkg/apis/aws/types_controlplane.go [hunk 2]
```diff
@@ -26,10 +26,21 @@ type ControlPlaneConfig struct {
...
 	// FeatureGates contains information about enabled feature gates.
 	FeatureGates map[string]bool
 }
+
+type Storage struct {
+	// ManagedDefaultClass controls if the 'default' StorageClass and 'default' VolumeSnapshotClass
+	// would be marked as default. Set to false to manually set the default to another class not
+	// managed by Gardener.
+	// Defaults to true.
+	ManagedDefaultClass *bool `json:"managedDefaultClass,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/aws/types_controlplane.go:45]:
```suggestion
	ManagedDefaultClass *bool
```

**Review by rfranzke — COMMENTED** [review]
Great, just small nits

File: pkg/controller/controlplane/valuesprovider_test.go [hunk 3]
```diff
@@ -269,13 +308,95 @@ var _ = Describe("ValuesProvider", func() {
...
+			}))
+		})
+
+		It("should have managedDefaultClass to true when using internal resource", func() {
+			internal := `{"kind":"ControlPlaneConfig","apiVersion":"aws.provider.extensions.gardener.cloud/__internal"}`
+
+			cp.Spec.DefaultSpec.ProviderConfig.Raw = []byte(internal)
+
+			x := string(cp.Spec.DefaultSpec.ProviderConfig.Raw)
+			fmt.Println(x)
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider_test.go:392]:
```suggestion
```

File: pkg/controller/controlplane/valuesprovider_test.go [hunk 4]
```diff
@@ -269,13 +308,95 @@ var _ = Describe("ValuesProvider", func() {
...
+			Expect(values).To(Equal(map[string]interface{}{
+				"useLegacyProvisioner": false,
+				"managedDefaultClass":  false,
+			}))
+		})
+
+		It("should have managedDefaultClass to true when using internal resource", func() {
+			internal := `{"kind":"ControlPlaneConfig","apiVersion":"aws.provider.extensions.gardener.cloud/__internal"}`
+
+			cp.Spec.DefaultSpec.ProviderConfig.Raw = []byte(internal)
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider_test.go:389]: Can't this be written also as

```suggestion
			cp.Spec.DefaultSpec.ProviderConfig.Raw = encode(&apisaws.ControlPlaneConfig{})
```

File: pkg/controller/controlplane/valuesprovider_test.go [hunk 5]
```diff
@@ -15,11 +15,15 @@
...
+	"bytes"
 	"context"
-	"encoding/json"
+	"fmt"
+	"io/ioutil"
 
 	apisaws "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws"
+	apisawsv1alpha1 "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/v1alpha1"
 	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
+	"github.com/go-logr/logr"
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider_test.go:26]: You could move it in the group below.

File: pkg/controller/controlplane/valuesprovider.go [hunk 6]
```diff
@@ -371,16 +371,35 @@ func (vp *valuesProvider) GetControlPlaneShootChartValues(
...
+	cp *extensionsv1alpha1.ControlPlane,
 	cluster *extensionscontroller.Cluster,
 ) (map[string]interface{}, error) {
 	k8sVersionLessThan118, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", "1.18")
 	if err != nil {
 		return nil, err
 	}
 
+	// Decode providerConfig
+	cpConfig := &apisaws.ControlPlaneConfig{}
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider.go:383]: You could move this line in before L387. It is used only in the if statement, so there is no need to expose it in the parent scope.

```suggestion
	if cp.Spec.ProviderConfig != nil {
		cpConfig := &apisaws.ControlPlaneConfig{}
		_, _, err := vp.Decoder().Decode(cp.Spec.ProviderConfig.Raw, nil, cpConfig)
```

File: pkg/apis/aws/v1alpha1/types_controlplane.go [hunk 7]
```diff
@@ -36,3 +40,12 @@ type CloudControllerManagerConfig struct {
 	// +optional
 	FeatureGates map[string]bool `json:"featureGates,omitempty"`
 }
+
+type Storage struct {
```
Comment (ialidzhikov) [inline pkg/apis/aws/v1alpha1/types_controlplane.go:44]: You could add a doc string.

File: pkg/apis/aws/types_controlplane.go [hunk 8]
```diff
@@ -26,10 +26,21 @@ type ControlPlaneConfig struct {
...
+	Storage *Storage
 }
 
 // CloudControllerManagerConfig contains configuration settings for the cloud-controller-manager.
 type CloudControllerManagerConfig struct {
 	// FeatureGates contains information about enabled feature gates.
 	FeatureGates map[string]bool
 }
+
+type Storage struct {
```
Comment (ialidzhikov) [inline pkg/apis/aws/types_controlplane.go:40]: You could add a doc string.

**Review by ialidzhikov — COMMENTED** [review]
Looks good in general. Minor nits

File: pkg/controller/controlplane/valuesprovider_test.go (hunk 4 above)
Comment (mvladev) [reply pkg/controller/controlplane/valuesprovider_test.go:389]: Encoder chooses external version over internal for the encoding. This won't work.

**Review by rfranzke — APPROVED** [review]

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**